File: server/agents/proactive_definer_agent.py
from langchain.prompts import PromptTemplate
from langchain.schema import HumanMessage
from langchain.chat_models import ChatOpenAI
from langchain.output_parsers import PydanticOutputParser
from langchain.schema import OutputParserException
from pydantic import BaseModel, Field
from agents.agent_utils import format_list_data
from server_config import openai_api_key
from agents.search_tool_for_agents import (
    asearch_google_knowledge_graph,
    search_url_for_entity_async,
)
from Modules.LangchainSetup import *
import asyncio

proactive_rare_word_agent_prompt_blueprint = """
# Objective
Your role is to identify and define "Rare Entities (REs)" in a transcript. Types of REs include rare words, jargons, adages, concepts, people, places, organizations, events etc that are not well known to the average high schooler, in accordance to current trends. You can also intelligently detect REs that are described in the conversation but not explicitly mentioned.

# Criteria for Rare Entities in order of importance
1. Rarity: Select entities that are unlikely for an average high schooler to know. Well known entities are like Fortune 500 organizations, worldwide-known events, popular locations, and entities popularized by recent news or events such as "COVID-19", "Bitcoin", or "Generative AI".
2. Utility: Definition should help a user understand the conversation better and achieve their goals.
3. No Redundancy: Exclude definitions if already defined in the conversation.
4. Complexity: Choose phrases with non-obvious meanings, such that their meaning cannot be derived from simple words within the entity name, such as "Butterfly Effect" which has a totally different meaning from its base words, but not "Electric Car" nor "Lane Keeping System" as they're easily derived.
5. Definability: Must be clearly and succinctly definable in under 10 words.
6. Existance: Don't select entities if you have no knowledge of them

# Conversation Transcript:
<Transcript start>{conversation_context}<Transcript end>

# Output Guidelines:
Output an array of the entities using the following template: `[{{ name: string, definition: string, search_keyword: string }}]`
- name is the RE name shown to the user, if the name is mistranscribed, autocorrect it into the most well known form with proper spelling, capitalization and punctuation
- definition is concise (< 12 words) and uses simple words
- search_keyword is the best specific Internet search keywords to search for the RE, you might need to use their complete official RE name, or autocorrect RE name, or add additional context keywords (like the entity type) for better searchability, especially if the entity is ambiguous or has multiple meanings. Additionally, for rare words, add "definition" to the search keyword.
- it's OK to output an empty array - most of the time, the array will be empty, only include items if the fit all the requirements

## Additional Guidelines:
- Only define NOUNS.
- Select RE that have an entry in an encyclopedia, wikipedia, dictionary, or other reference material.
- Do not select a RE you yourself are unfamiliar with, you can infer the implied entity or autocorrect mistransribed names only if you are 99% confident, never select an entity if you will define it as "Unknown Entity".
- Multiple REs can be defined from one phrase, for example, "The Lugubrious Game" is a candidate to define, the rare word "lugubrious" is also a candidate.
- Limit results to {number_of_definitions} REs.
- Examples:
    - Completing incomplete name example: Conversation mentions "Balmer" and "Microsoft", the keyword is "Steve Balmer + person", and the name would be "Steve Balmer" because it is complete.
    - Replacing unofficial name example: Conversation mentions "Clay Institute", the keyword is "Clay Mathematics Institute + organization", using the official name.
    - Add context example: Conversation mentions "Theory of everything", the keyword needs context keywords such as "Theory of everything + concept", because there is a popular movie with the same name. 
    - Autocorrect transcript example: Conversation mentions "Coleman Sachs" in the context of finance, if you are confident it is supposed to be "Goldman Sachs", you autocorrect it and define "Goldman Sachs".

## Recent Definitions:
These REs have already been defined so don't define them again:
{definitions_history}

## Example Output:
entities: [{{ name: "80/20 Rule", definition: "Productivity concept; Majority of results come from few causes", search_keyword: "80/20 Rule + concept" }}]

{format_instructions} 
If there are no relevant entities, output an empty array.
"""
# 6. Searchability: Likely to have a specific and valid reference source: Wikipedia page, dictionary entry etc.
# - Entity names should be quoted from the conversation, so the output definitions can be referenced back to the conversation.


class Entity(BaseModel):
    name: str = Field(
        description="entity name",
    )
    definition: str = Field(
        description="entity definition",
    )
    search_keyword: str = Field(
        description="keyword to search for entity on the Internet",
    )


class ConversationEntities(BaseModel):
    entities: list[Entity] = Field(
        description="list of entities and their definitions", default=[]
    )


proactive_rare_word_agent_query_parser = PydanticOutputParser(
    pydantic_object=ConversationEntities
)


def run_proactive_definer_agent(
    conversation_context: str, definitions_history: list = []
):
    # start up GPT4 connection
    llm = get_langchain_gpt4()

    extract_proactive_rare_word_agent_query_prompt = PromptTemplate(
        template=proactive_rare_word_agent_prompt_blueprint,
        input_variables=[
            "conversation_context",
            "definitions_history",
        ],
        partial_variables={
            "format_instructions": proactive_rare_word_agent_query_parser.get_format_instructions(),
            "number_of_definitions": "3",  # this is a tradeoff between speed and results, 3 is faster than 5
        },
    )

    if len(definitions_history) > 0:
        definitions_history = format_list_data(definitions_history)
    else:
        definitions_history = "None"

    proactive_rare_word_agent_query_prompt_string = (
        extract_proactive_rare_word_agent_query_prompt.format_prompt(
            conversation_context=conversation_context,
            definitions_history=definitions_history,
        ).to_string()
    )

    # print("Proactive meta agent query prompt string", proactive_rare_word_agent_query_prompt_string)

    response = llm(
        [HumanMessage(content=proactive_rare_word_agent_query_prompt_string)]
    )

    print("Proactive meta agent response", response)

    try:
        res = proactive_rare_word_agent_query_parser.parse(response.content)
        # we still have unknown_entities to search for but we will do them next time
        res = search_entities(res.entities)
        return res
    except OutputParserException:
        return None


def search_entities(entities: list[Entity]):
    search_tasks = []
    for entity in entities:
        search_tasks.append(search_url_for_entity_async(entity.search_keyword))

    loop = asyncio.get_event_loop()
    responses = asyncio.gather(*search_tasks)
    responses = loop.run_until_complete(responses)

    entity_objs = []
    for entity, response in zip(entities, responses):
        # print("response", str(response))
        res = dict()
        res["name"] = entity.name
        res["summary"] = entity.definition
        res.update(response)

        # if response is None:
        #     continue

        # for item in response.item_list_element:
        #     result = item.get("result")

        #     # get mid and start entry - assuming we always get a mid
        #     mid = None
        #     for identifier in result.get("identifier"):
        #         if identifier.get('name') == 'googleKgMID':
        #             mid = identifier.get('value')
        #             break

        #     if mid is None:
        #         continue

        #     res["mid"] = mid

        #     # get google cloud id
        #     cloud_id = result.get('@id')

        #     # get image
        #     if result.get('image'):
        #         image_url = result.get('image').get('contentUrl')
        #         # convert to actual image url if it's a wikipedia image
        #         # if "wiki" in image_url:
        #         # image_url = self.wiki_image_parser(image_url)
        #         res["image_url"] = image_url

        #     # res["name"] = result.get('name')
        #     res["category"] = result.get('description')

        #     # set our own types
        #     if any(x in ['Place', 'City', 'AdministrativeArea'] for x in result.get('@type')):
        #         res["type"] = "LOCATION"
        #     else:
        #         res["type"] = result.get('@type')[0].upper()

        #     detailed_description = result.get("detailedDescription")
        #     if detailed_description:
        #         res["url"] = detailed_description.get('url')

        entity_objs.append(res)

    return entity_objs

File: augmentos_cloud/packages/agents/ProactiveDefinerAgent.ts
import { Agent } from './AgentInterface';
import { PromptTemplate } from '@langchain/core/prompts';
import { JsonOutputFunctionsParser } from 'langchain/output_parsers';
import { LLMProvider } from '../utils/LLMProvider';
import { SearchToolForAgents } from './tools/SearchToolForAgents';

// ===== Type Definitions =====

export interface Entity {
  name: string;
  definition: string;
  search_keyword: string;
}

export interface ConversationEntities {
  entities: Entity[];
  irrelevant_terms: string[];
}

// ===== Prompt Blueprint =====

const proactiveRareWordAgentPromptBlueprint = `# Objective
Your role is to identify and define "Rare Entities (REs)" in a transcript. Types of REs include rare words, jargons, adages, concepts, people, places, organizations, events etc that are not well known to the average high schooler, in accordance to current trends. You can also intelligently detect REs that are described in the conversation but not explicitly mentioned.

# Criteria for Rare Entities in order of importance
1. **Rarity:** Select entities that are unlikely for an average high schooler to know. Do not include well-known entities.
2. **Utility:** The definition should help a user understand the conversation better and achieve their goals.
3. **No Redundancy:** Exclude definitions if they are already defined in the conversation.
4. **Complexity:** Choose phrases whose meaning is not obvious from the component words.
5. **Definability:** Each entity must be clearly and succinctly definable in under 10 words.
6. **Existence:** Do not select an entity if you have no reference knowledge of it.

# Conversation Transcript:
<Transcript start>{conversation_context}<Transcript end>

# Additional Context:
Existing definitions history: {definitions_history}

# Output Guidelines:
Output an array \`entities\` of the entities using the following template:
\`\`\`
[
  { name: string, definition: string, search_keyword: string }
]
\`\`\`
Also output an array \`irrelevant_terms\` of terms that are not rare or interesting.

Example Output:
entities: [{ name: "80/20 Rule", definition: "Productivity concept; Majority of results come from few causes", search_keyword: "80/20 Rule + concept" }]
irrelevant_terms: ["the", "dog", "rock", "table", "who"]

{format_instructions}`;

// Create a parser for the expected JSON output
const proactiveRareWordAgentQueryParser = new JsonOutputFunctionsParser<ConversationEntities>({ argsOnly: true });

/**
 * The ProactiveDefinerAgent identifies and defines rare entities from a conversation transcript.
 */
export class ProactiveDefinerAgent implements Agent {
  public agentId = 'proactive_definer';
  public agentName = 'Proactive Definer';
  public agentDescription = 'Identifies and defines rare entities in a conversation transcript.';
  public agentPrompt = proactiveRareWordAgentPromptBlueprint;
  public agentTools = ['search'];

  /**
   * Expects an input object with the following keys:
   * - conversation_context: string
   * - definitions_history?: string[]
   * - irrelevant_terms?: string[]
   */
  public async handleContext(inputData: Record<string, any>): Promise<{ [key: string]: any }> {
    const conversation_context: string = inputData.conversation_context;
    const definitions_history: string[] = inputData.definitions_history || [];
    const irrelevant_terms: string[] = inputData.irrelevant_terms || [];

    // Format definitions_history for the prompt
    const formattedDefinitionsHistory = definitions_history.length > 0 ? JSON.stringify(definitions_history) : "None";

    // Build the prompt template
    const promptTemplate = new PromptTemplate({
      template: this.agentPrompt,
      inputVariables: ['conversation_context', 'definitions_history'],
      partialVariables: {
        format_instructions: proactiveRareWordAgentQueryParser.getFormatInstructions(),
        number_of_definitions: "1", // Adjust if needed
      },
    });

    // Format the final prompt string
    const formattedPrompt = await promptTemplate.format({
      conversation_context,
      definitions_history: formattedDefinitionsHistory,
    });

    // Get the LLM instance (e.g., GPT-4 or another supported model)
    const llm = LLMProvider.getLLM();

    try {
      // Invoke the LLM with the formatted prompt
      const response = await llm.invoke(formattedPrompt);

      // Parse the JSON output to match our ConversationEntities interface
      const parsed: ConversationEntities = await proactiveRareWordAgentQueryParser.parse(response.content);

      // Enrich each entity with additional search data
      const enrichedEntities = await this.searchEntities(parsed.entities);

      return {
        entities: enrichedEntities,
        irrelevant_terms: parsed.irrelevant_terms,
      };
    } catch (error) {
      console.error("Error in ProactiveDefinerAgent.handleContext:", error);
      return { entities: [], irrelevant_terms: [] };
    }
  }

  /**
   * Searches for additional information (and images if enabled) for a list of entities.
   *
   * @param entities - list of entities to enrich.
   * @returns A Promise resolving to an array of enriched entity objects.
   */
  private async searchEntities(entities: Entity[]): Promise<any[]> {
    // Create search tasks for each entity
    const searchTasks = entities.map((entity) =>
      searchUrlForEntityAsync(entity.search_keyword, false) // Adjust if image data is needed
    );

    const responses = await Promise.all(searchTasks);

    // Combine the LLM-provided data with the search result
    const enrichedEntities = entities.map((entity, index) => {
      const response = responses[index] || {};
      return {
        name: entity.name,
        summary: entity.definition,
        ...response,
      };
    });

    return enrichedEntities;
  }
}
